domain/application/service/application.go

```diff
@@ -1245,6 +1245,71 @@ func (s *Service) GetApplicationDisplayStatus(ctx context.Context, appID coreapp
 	return derivedApplicationStatus, nil
 }
 
+// GetApplicationAndUnitStatusesForUnitWithLeader returns the display status
+// of the application the specified unit belongs to, and the workload statuses
+// of all the units that belong to that application, indexed by unit name.
+// If no application is found for the unit name, an error satisfying
+// [applicationerrors.ApplicationNotFound] is returned
+func (s *Service) GetApplicationAndUnitStatusesForUnitWithLeader(
+	ctx context.Context,
+	unitName coreunit.Name,
+) (
+	*corestatus.StatusInfo,
+	map[coreunit.Name]corestatus.StatusInfo,
+	error,
+) {
+	if err := unitName.Validate(); err != nil {
+		return nil, nil, internalerrors.Errorf("unit name: %w", err)
+	}
+
+	appName := unitName.Application()
+
+	var applicationDisplayStatus *corestatus.StatusInfo
+	var unitWorkloadStatuses map[coreunit.Name]corestatus.StatusInfo
+
+	err := s.leaderEnsurer.WithLeader(ctx, appName, unitName.String(), func(ctx context.Context) error {
+		appID, err := s.st.GetApplicationIDByName(ctx, appName)
+		if err != nil {
+			return internalerrors.Errorf("getting application id: %w", err)
+		}
```

Comment:
This should be done before the `WithLeader` check. It's cheaper to check if that doesn't exist before the leader checks.

Reply:
Fair. Done